shared/modules/transaction.utils.ts

```diff
@@ -188,36 +179,34 @@ export async function determineTransactionType(txParams, query) {
   return { type: result, getCodeResponse: contractCode };
 }
 
-const INFERRABLE_TRANSACTION_TYPES = [
-  TransactionType.tokenMethodApprove,
-  TransactionType.tokenMethodSetApprovalForAll,
-  TransactionType.tokenMethodTransfer,
-  TransactionType.tokenMethodTransferFrom,
-  TransactionType.contractInteraction,
-  TransactionType.simpleSend,
-];
-
+type GetTokenStandardAndDetails = (to: string | undefined) => {
+  decimals?: string;
+  balance?: string;
+  symbol?: string;
+  standard?: TokenStandard;
+};
 /**
  * Given a transaction meta object, determine the asset type that the
  * transaction is dealing with, as well as the standard for the token if it
  * is a token transaction.
  *
- * @param {import('@metamask/transaction-controller').TransactionMeta} txMeta -
- *  transaction meta object
- * @param {EthQuery} query - EthQuery instance
- * @param {Function} getTokenStandardAndDetails - function to get token
- *  standards and details.
- * @returns {{ assetType: string, tokenStandard: string}}
+ * @param txMeta - transaction meta object
+ * @param query - EthQuery instance
+ * @param getTokenStandardAndDetails - function to get token standards and details.
+ * @returns assetType: AssetType, tokenStandard: TokenStandard
  */
 export async function determineTransactionAssetType(
-  txMeta,
-  query,
-  getTokenStandardAndDetails,
-) {
+  txMeta: TransactionMeta,
+  query: EthQuery,
+  getTokenStandardAndDetails: GetTokenStandardAndDetails,
+): Promise<{
+  assetType: AssetType;
+  tokenStandard: TokenStandard;
+}> {
   // If the transaction type is already one of the inferrable types, then we do
   // not need to re-establish the type.
   let inferrableType = txMeta.type;
-  if (INFERRABLE_TRANSACTION_TYPES.includes(txMeta.type) === false) {
+  if (!INFERRABLE_TRANSACTION_TYPES.includes(txMeta.type as TransactionType)) {
```

Comment:
Let's resolve this by letting the compiler narrow the type (+ adding a runtime check) instead of relying on a type assertion.

```suggestion
  if (txMeta.type && !INFERRABLE_TRANSACTION_TYPES.includes(txMeta.type)) {
```

A null check or non-nullable assertion removes the `| undefined`, `| null` from a type i.e. it's a negative definition, enforcing what the type is NOT. A type assertion enforces what the type IS, and is therefore more dangerous.

For example, if we defined a `TransactionMetaType` somewhere down the road, the `as TransactionType` assertion would prevent the compiler from inferring that type, while a null check or non-nullable assertion would not.

Reply:
Thanks for the advice, will add !